File: clients/dart/tool/tjsv_rpc_probe.dart
// Fixed stdin/stdout conformance adapter; no production options or parser.
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import '../lib/rpc_v1.dart';

const _limit = 16 * 1024 * 1024;

void _require(bool condition, String message) {
  if (!condition) throw StateError(message);
}

Map<String, Object?> _evaluate(Map<String, dynamic> row) {
  final name = row['name'] as String;
  final kind = row['kind'] as String;
  final encoded = row['encoded'] as String;
  Object value;
  try {
    value = switch (kind) {
      'call' => decodeRpcV1Call(encoded),
      'receipt' => decodeRpcV1Receipt(encoded),
      _ => throw StateError('unknown probe kind'),
    };
  } on RpcV1Exception {
    return {'name': name, 'kind': kind, 'accepted': false};
  }
  // Only decoder exceptions count as rejection. Encoder failures and crashes
  // propagate to a failing process, never a successful negative observation.
  final bytes = switch (value) {
    RpcV1Call() => encodeRpcV1Call(value),
    RpcV1Receipt() => encodeRpcV1Receipt(value),
    _ => throw StateError('unexpected decoded type'),
  };
  return {
    'name': name,
    'kind': kind,
    'accepted': true,
    'encoded': utf8.decode(bytes),
  };
}

Future<void> main(List<String> args) async {
  try {
    _require(args.isEmpty, 'probe accepts no arguments');
    final input = BytesBuilder(copy: false);
    await for (final chunk in stdin) {
      _require(input.length + chunk.length <= _limit, 'probe input exceeds limit');
      input.add(chunk);
    }
    final document = jsonDecode(utf8.decode(input.takeBytes()));
    _require(document is Map<String, dynamic>, 'probe must be an object');
    final request = document as Map<String, dynamic>;
    _require(
      request.length == 2 &&
          request['schema'] == 'ores.api-docs.rpc-probe/v1' &&
          request['cases'] is List,
      'invalid probe schema or fields',
    );
    final cases = request['cases'] as List;
    _require(cases.isNotEmpty && cases.length <= 4096, 'invalid probe coverage');
    final names = <String>{};
    final results = <Map<String, Object?>>[];
    for (final value in cases) {
      _require(value is Map<String, dynamic>, 'case must be an object');
      final row = value as Map<String, dynamic>;
      _require(
        row.length == 3 &&
            row['name'] is String &&
            row['kind'] is String &&
            row['encoded'] is String,
        'invalid probe case fields',
      );
      final name = row['name'] as String;
      _require(name.isNotEmpty && names.add(name), 'empty or duplicate probe name');
      results.add(_evaluate(row));
    }
    stdout.writeln(jsonEncode({
      'schema': 'ores.api-docs.rpc-probe-result/v1',
      'runtime': 'dart',
      'results': results,
    }));
    await stdout.flush();
  } catch (error) {
    stderr.writeln('probe execution failed: $error');
    exitCode = 3;
  }
}

File: clients/rust/examples/tjsv_rpc_probe.rs
//! Fixed stdin/stdout conformance adapter, not a production CLI or validator.
use ores_api_docs_client::{decode_rpc_v1_call, decode_rpc_v1_receipt};
use serde_json::{json, Value};
use std::collections::HashSet;
use std::error::Error;
use std::io::{self, Read, Write};

const LIMIT: u64 = 16 * 1024 * 1024;

fn text<'a>(value: &'a Value, field: &str) -> Result<&'a str, Box<dyn Error>> {
    value.get(field).and_then(Value::as_str).ok_or_else(|| "invalid probe field".into())
}

fn run() -> Result<(), Box<dyn Error>> {
    if std::env::args_os().len() != 1 {
        return Err("probe accepts no arguments".into());
    }
    let mut input = Vec::new();
    io::stdin().lock().take(LIMIT + 1).read_to_end(&mut input)?;
    if input.len() as u64 > LIMIT {
        return Err("probe input exceeds limit".into());
    }
    let document: Value = serde_json::from_slice(&input)?;
    let fields = document.as_object().ok_or("probe must be an object")?;
    if fields.len() != 2 || text(&document, "schema")? != "ores.api-docs.rpc-probe/v1" {
        return Err("invalid probe schema or fields".into());
    }
    let cases = document.get("cases").and_then(Value::as_array).ok_or("missing probe cases")?;
    if cases.is_empty() || cases.len() > 4096 {
        return Err("invalid probe case count".into());
    }
    let mut names = HashSet::new();
    let mut results = Vec::with_capacity(cases.len());
    for case in cases {
        if case.as_object().ok_or("case must be an object")?.len() != 3 {
            return Err("unexpected probe case fields".into());
        }
        let name = text(case, "name")?;
        let kind = text(case, "kind")?;
        let encoded = text(case, "encoded")?;
        if name.is_empty() || !names.insert(name) {
            return Err("empty or duplicate probe name".into());
        }
        // Only the real decoder's Result::Err is rejection. Encoder and IO errors
        // propagate as process failure, and a panic is never recovered as evidence.
        let encoded_result = match kind {
            "call" => match decode_rpc_v1_call(encoded.as_bytes()) {
                Ok(value) => Some(value.encode()?),
                Err(_) => None,
            },
            "receipt" => match decode_rpc_v1_receipt(encoded.as_bytes()) {
                Ok(value) => Some(value.encode()?),
                Err(_) => None,
            },
            _ => return Err("unknown probe kind".into()),
        };
        results.push(match encoded_result {
            Some(bytes) => json!({"name": name, "kind": kind, "accepted": true, "encoded": String::from_utf8(bytes)?}),
            None => json!({"name": name, "kind": kind, "accepted": false}),
        });
    }
    let output = json!({"schema": "ores.api-docs.rpc-probe-result/v1", "runtime": "rust", "results": results});
    let mut stdout = io::stdout().lock();
    serde_json::to_writer(&mut stdout, &output)?;
    writeln!(stdout)?;
    Ok(())
}

fn main() {
    if let Err(error) = run() {
        eprintln!("probe execution failed: {error}");
        std::process::exit(3);
    }
}
